### ISSUE TITLE
How to add filters depending on database values?

### ISSUE THREAD

**Issue by jwillmer** [jcachat/EntityFramework.DynamicFilters#30]
Is it possible to query the databse before setting up the filters? I like to get a companyId from the database to filter all other tables with that id.

**Comment by siegeon** [comment]
You can disable all filters at any time, then re-enable your filters.

using (var context = new TestContext())
			{
				context.DisableAllFilters();

				var companyId = GetCompanyId();

                                context.EnableAllFilters();
			}

**Comment by jcachat** [comment]
You can create your filters before you actually set the value to filter on.  Take a look at how the "BlogEntryFilter" filter is configured in ExampleContext.cs of the Example project.  It is creating a filter against a static property "CurrentAccountID".  The value for that property just needs to be set before you actually execute a query that would invoke the filter (otherwise, in this case, it would use the default value for a Guid).

**Comment by siegeon** [comment]
jcachat If you did not disable the filter before running the query wouldn't it filter all the results by guid.default?

**Comment by jcachat** [comment]
No, the value to filter on isn't read until the filter is actually executed - and it's read each time it's executed (and set into a sql parameter).  That is what makes it dynamic vs. the EntityFramework.Filters project which reads the values when the query is COMPILED.  So when you use a delegate to get the value of a filter parameter (as in this example), you can freely change it any time you want and the filter will always use the current value each time it executes.

Disabling the filter actually just sets a bool property value on the query that causes the conditions (which have already been compiled into the query) to not be applied at query execution time.  If you use SQL Profiler to capture the sql when the filter is disabled vs. enabled, you'll see that is the only difference is in the value of a sql parameter named like "@DynamicFilterParam_[filtername]_DynamicFilterIsDisable".

**Comment by siegeon** [comment]
Let me rephrase, and please correct me if I am wrong. Looking at the OP I understood that he wanted to query the database for the Company Id before the filter that needed the Id is set up. In that instance the first query against the databases to get the Company Id would be filtered by Guid.Empty when the query is executed the first time.

**Comment by jcachat** [comment]
Oh I see what you mean now.  I was assuming that the query to get the Company ID is being done against something that is not filtered (some kind of configuration table that contains the Company ID) while the filters are set on entities that have a CompanyID column/FK.

It looks like your first post is much more detailed than the email git sent me so I didn't see all that (maybe you edited).  So in your example, if your call to:
```
CompanyRequestManager.Instance.CompanyId = context.CompanyEntity.GetIdSomehow()
```
does a query against an entity that is ICompanyOwned, then yes, you need to disable the filter first.  If not, you do not need to disable.

That is a good point to make but otherwise it looks like we answered the original question the same way.

**Comment by siegeon** [comment]
Yeah I totally edited my first response after re-reading the OP. =)

**Comment by jwillmer** [comment]
Thanks for the fast response! Your code example and the discussion helped me a lot :-)

**Comment by jwillmer** [comment]
I created a singelton class that queries for the company id but I get an exception.
    
 > The context cannot be used while the model is being created. This exception may be thrown if the context is used inside the OnModelCreating method..

I understand the problem, the company id is instantly retrieved:
  
    modelBuilder.Filter("GlobalCompanyFilter", (ICompanyReference r) => r.CompanyId, SingletonPerRequest.Current.CompanyId);

And another problem is that disabling the filter wouldn't stop retrieving the value because as @jcachat said: 

 > Disabling the filter actually just sets a bool property value on the query that causes the conditions (which have already been compiled into the query) to not be applied at query execution time.

@siegeon how did you resolve this problem?

**Comment by jwillmer** [comment]
I think I solved it:

    modelBuilder.Filter("GlobalCompanyFilter", (ICompanyReference r) => r.CompanyId, () => SingletonPerRequest.Current.CompanyId);
